[Fix-Improve] Improve Ollama prompt input and fix Ollama function calling key error and fix Ollama function calling `can only join an iterable` error

corrm opened the pull request:
## Title

Improve Ollama prompt input

## Relevant issues

Fixes #1272
Fixes #3912
Fixes(Kinda) https://github.com/microsoft/autogen/issues/2924
RP #4089

## Type

🐛 Bug Fix
🧹 Refactoring

## Changes

- Resolve the 'name' key error by use `dict.get` method and fallback to the 'function' key if not present
- Improve Ollama prompt input
- Fix "litellm.exceptions.APIConnectionError: can only join an iterable"

## [REQUIRED] Testing - Attach a screenshot of any new tests passing local

Ollama requires local testing
testing the example from https://microsoft.github.io/autogen/docs/topics/non-openai-models/local-litellm-ollama/

program response
```
user_proxy (to chatbot):

How much is 123.45 EUR in USD?

--------------------------------------------------------------------------------
chatbot (to user_proxy):

***** Suggested tool call (call_b2eb58bd-5ff4-4332-9566-147fb17d4b76): currency_calculator *****
Arguments: 
{"base_amount": 123.45, "quote_currency": "USD"}
************************************************************************************************

--------------------------------------------------------------------------------

>>>>>>>> EXECUTING FUNCTION currency_calculator...
user_proxy (to chatbot):

user_proxy (to chatbot):

***** Response from calling tool (call_b2eb58bd-5ff4-4332-9566-147fb17d4b76) *****
123.45 USD
**********************************************************************************

--------------------------------------------------------------------------------
chatbot (to user_proxy):

***** Suggested tool call (call_8d3ca275-0b00-4bd1-8734-3f2d9f1e438a): currency_calculator *****
Arguments: 
{"parameter_1_name": "base_amount", "parameter_2_name": "quote_currency"}
************************************************************************************************

--------------------------------------------------------------------------------

Process finished with exit code 0
```

corrm committed:
chore: Improved prompt generation in ollama_pt function

corrm committed:
chore: Improved OllamaConfig get_required_params and ollama_acompletion and ollama_async_streaming functions

corrm committed:
Added improved function name handling in ollama_async_streaming

corrm committed:
Rename ollama prompt 'Function' word to 'Name'

krrishdholakia commented:
Hey @CorrM can you share a screenshot of it passing this - https://github.com/BerriAI/litellm/blob/95f972ee9fee68992b5932fe921590f0daa67247/litellm/tests/test_ollama.py

```
cd litellm/litellm

pytest test_ollama.py -x -vv
```

corrm committed:
Rename ollama prompt:
- 'Function' word to 'FunctionName'
- 'Tool Call' to `FunctionCall`
- 'Tool Call Result' to 'FunctionCall Result'

_I found that changes make some models better_

CorrM commented:
> Hey @CorrM can you share a screenshot of it passing this - https://github.com/BerriAI/litellm/blob/95f972ee9fee68992b5932fe921590f0daa67247/litellm/tests/test_ollama.py
> 
> ```
> cd litellm/litellm
> 
> pytest test_ollama.py -x -vv
> ```

![image](https://github.com/BerriAI/litellm/assets/5868825/b24a0d45-a0e8-4736-b8e5-4a35f4c29485)

corrm committed:
Improve ollama prompt: this formula give good result with AutoGen

CorrM commented:
Same test results after https://github.com/BerriAI/litellm/pull/4373/commits/5b720137fdb2e65c9584bc1dbdb7a7e50120eb7e commit

```
test_ollama.py::test_get_ollama_model PASSED                                                                                                                                                                                [ 20%]
test_ollama.py::test_get_ollama_params PASSED                                                                                                                                                                               [ 40%]
test_ollama.py::test_ollama_aembeddings PASSED                                                                                                                                                                              [ 60%]
test_ollama.py::test_ollama_embeddings PASSED                                                                                                                                                                               [ 80%]
test_ollama.py::test_ollama_json_mode PASSED                                                                                                                                                                                [100%]
```
